xarray/core/combine.py

```diff
@@ -207,8 +207,29 @@ def _dataset_concat(datasets, dim, data_vars, coords, compat, positions):
 
     dim, coord = _calc_concat_dim_coord(dim)
     datasets = [as_dataset(ds) for ds in datasets]
+
+    # GH1434
+    # constructing a dictionary that will be used to preserve dtype
+    # of the original dataset dimensions
+    dtype_dict = {}
+    for ds in datasets:
+        for dim_tuple in ds.dims.items():
+            dim_name = dim_tuple[0]
+            if dim_name != dim:
+                dtype_dict[dim_name] = ds[dim_name].dtype
+
+    # align loses original dtype of the datasets' dim variables
     datasets = align(*datasets, join='outer', copy=False, exclude=[dim])
 
+    # GH1434
+    # restoring original dtype of the datasets' dimensions
```

Comment:
This fix should either be done in `align` (actually inside `core.alignment.reindex_variables`) or not at all. We cannot preserve all dtypes in alignment operations, e.g., there is no missing value for structured dtypes. If you care about this, you should use `join='inner'` or `join='exact'` (will appears in the next release, from https://github.com/pydata/xarray/pull/1330). It would be nice to surface these options up into `xarray.concat`, though.